libpod/oci_conmon_linux.go

```diff
@@ -40,35 +42,63 @@ func (r *ConmonOCIRuntime) createRootlessContainer(ctr *Container, restoreOption
 			}
 			defer errorhandling.CloseQuiet(fd)
 
+			rootPath, err := ctr.getRootPathForOCI()
+			if err != nil {
+				return 0, err
+			}
+
 			// create a new mountns on the current thread
 			if err = unix.Unshare(unix.CLONE_NEWNS); err != nil {
 				return 0, err
 			}
 			defer func() {
-				if err := unix.Setns(int(fd.Fd()), unix.CLONE_NEWNS); err != nil {
-					logrus.Errorf("Unable to clone new namespace: %q", err)
+				err := unix.Setns(int(fd.Fd()), unix.CLONE_NEWNS)
+				if err == nil {
+					// If we are able to reset the previous mount namespace, unlock the thread and reuse it
+					runtime.UnlockOSThread()
+				} else {
+					// otherwise, leave the thread locked and the Go runtime will terminate it
+					logrus.Errorf("Unable to reset the previous mount namespace: %q", err)
 				}
 			}()
 
-			// don't spread our mounts around.  We are setting only /sys to be slave
-			// so that the cleanup process is still able to umount the storage and the
-			// changes are propagated to the host.
-			err = unix.Mount("/sys", "/sys", "none", unix.MS_REC|unix.MS_SLAVE, "")
-			if err != nil {
-				return 0, fmt.Errorf("cannot make /sys slave: %w", err)
+			if rootPath != "" {
+				parentDir := filepath.Dir(rootPath)
+				if err := unix.Mount(parentDir, parentDir, "", unix.MS_BIND, ""); err != nil {
+					return 0, fmt.Errorf("making intermediate parent directory for container %s: %w", ctr.ID(), err)
+				}
+				if err := unix.Mount("", parentDir, "", unix.MS_PRIVATE, ""); err != nil {
+					return 0, fmt.Errorf("making intermediate parent directory private for container %s: %w", ctr.ID(), err)
+				}
```

Comment:
If we make this mount private doesn't this mean a users can no longer receive additional mounts in their mounted volumes (assuming they specified the shared,slave option for the volume)

i.e. assume the container has /tmp/test mounted and then on the host someone mounts /tmp/test/mnt then it no longer gets forwarded? Or am I missing something? I have not tested this yet.

Reply:
no, this prevents only mounts below `parentDir` to not be propagated, everything else maintains their original propagation.

The PR was still marked Draft because I was not happy with the hack above, and it was still propagating a new mount to the parent mount namespace.  I've submitted a new version that doesn't require the double mount